proposals/sealed-inline-classes.md

```diff
@@ -0,0 +1,593 @@
+
+## Motivation / use cases
+
+The main use-case for sealed inline classes is `Result`. Currently it is declared as inline class with underlying type `Any?`
+```kotlin
+@JvmInline
+value class Result<out T>(val value: Any?) {
+  class Failure(val throwable: Throwable)
+}
+```
+
+The proposed changes to `Result` are
+```kotlin
+@JvmInline
+sealed value class Result<out T> {
+  @JvmInline
+  value class Success<out T>(val value: T): Result<T>()
+  value class Failure(val throwable: Throwable): Result<Nothing>()
+}
+```
+
+The changes also rely on [inline classes with generic underlying value](https://youtrack.jetbrains.com/issue/KT-32162/Allow-generics-for-inline-classes).
+
+So, to create the `Result` value one will use constructors instead of factory functions. In other words, instead of
+```kotlin
+val success = Result.success(1)
+val failure = Result.failure(IllegalStateException("fail"))
+```
+
+a programmer can write
+```kotlin
+val success = Result.Success(1)
+val failure = Result.Failure(IllegalStateException("fail"))
+```
+
+In addition, since sealed inline classes are sealed classes, one can use `when` to check the value, in case of `Result` 
+one currently should use `isSuccess` and `isFailure` utility function, another solution is, of course, `getOrThrow`.
+
+So,
+```kotlin
+val s = try {
+  result.getOrThrow()
+} catch(e: Throwable) {
+  e.message!!
+}
+```
+
+becomes
+```koltin
+when(result) {
+  is Result.Success<*> -> result.value
+  is Result.Failure -> result.throwable.message!!
+}
+```
+
+As you can see, sealed inline classes support smart casts.
+
+## Restrictions
+
+Sealed inline classes are both inline classes and sealed classes. Thus, they are implemented differently from usual inline classes.
+
+- Sealed inline class children can be either inline and noinline. For example, in `Result`, `Success` is an inline class 
+  and `Failure` - noinline. Inline children shall be annotated with `@JvmInline` annotation, since they are inline classes. 
+- In Kotlin/JS and Kotlin/Native, there is no need to annotate the inline children.
+
+- All sealed inline classes are mapped to `Any?`. That is because their underlying types are union of all the inline class children's underlying types with noinline children types. 
+
+For example, if the sealed inline class is declared as
+```kotlin
+@JvmInline
+sealed value class IC {
+  @JvmInline
+  value class ICString(val s: String): IC()
+  @JvmInline
+  value class ICInt(val i: Int): IC()
+  value class Error(val throwable: Throwable): IC()
+}
+```
+
+The resulting underlying type is `String | Int | IC.Error`, which is `Any?`. 
+
+Of course, we can map to common supertype. Consider the following example
+```kotlin
+@JvmInline
+sealed value class IC {
+  @JvmInline
+  value class ICInt(val i: Int): IC()
+  @JvmInline
+  value class ICLong(val l: Long): IC()
+}
+```
+
+here, we can use `Number` as the underlying type. However, adding non-`Number` child will change the underlying type, 
```

Comment:
It is better to replace "we can use Number" with "we could have used Number" to let a reader understand what's the actual design we are going with and what is just a digression on other potential alternatives we have considered in the design.